ompi/mca/coll/acoll/coll_acoll_utils.h

```diff
@@ -0,0 +1,882 @@
+            free(subgrp_ranks);
+            subgrp_ranks = NULL;
+        }
+        if (numa_ranks != NULL) {
+            free(numa_ranks);
+            numa_ranks = NULL;
+        }
+    }
+
+    /* Restore originals for local and socket comms */
+    (subc->local_comm)->c_coll->coll_allreduce = coll_allreduce_loc;
+    (subc->local_comm)->c_coll->coll_allgather = coll_allgather_loc;
+    (subc->local_comm)->c_coll->coll_bcast = coll_bcast_loc;
+    (subc->socket_comm)->c_coll->coll_allreduce = coll_allreduce_soc;
+    (subc->socket_comm)->c_coll->coll_allgather = coll_allgather_soc;
+    (subc->socket_comm)->c_coll->coll_bcast = coll_bcast_soc;
+
+    /* For collectives where order is important (like gather, allgather),
+     * split based on ranks. This is optimal for global communicators with
+     * equal split among nodes, but suboptimal for other cases.
+     */
+    if (!subc->initialized) {
+        if (subc->num_nodes > 1) {
+            int node_size = (size + subc->num_nodes - 1) / subc->num_nodes;
+            int color = rank / node_size;
+            err = ompi_comm_split(comm, color, rank, &subc->local_r_comm, false);
+            if (MPI_SUCCESS != err) {
+                return err;
+            }
+        }
+        subc->derived_node_size = (size + subc->num_nodes - 1) / subc->num_nodes;
+    }
+
+    /* Restore originals */
+    (comm)->c_coll->coll_allreduce = coll_allreduce_org;
+    (comm)->c_coll->coll_allgather = coll_allgather_org;
+    (comm)->c_coll->coll_bcast = coll_bcast_org;
+
+    /* Init done */
+    subc->initialized = 1;
+    if (root != subc->prev_init_root) {
+        subc->num_root_change++;
+    }
+    subc->prev_init_root = root;
+
+    return err;
+}
+
+#ifdef HAVE_XPMEM_H
+static inline int mca_coll_acoll_xpmem_register(void *xpmem_apid, void *base, size_t size,
+                                                mca_rcache_base_registration_t *reg)
+{
+    struct xpmem_addr xpmem_addr;
+    xpmem_addr.apid = *((xpmem_apid_t *) xpmem_apid);
+    xpmem_addr.offset = (uintptr_t) base;
+    struct acoll_xpmem_rcache_reg_t *xpmem_reg = (struct acoll_xpmem_rcache_reg_t *) reg;
+    xpmem_reg->xpmem_vaddr = xpmem_attach(xpmem_addr, size, NULL);
+
+    if ((void *) -1 == xpmem_reg->xpmem_vaddr) {
+        return -1;
+    }
+    return 0;
+}
+
+static inline int mca_coll_acoll_xpmem_deregister(void *xpmem_apid,
+                                                  mca_rcache_base_registration_t *reg)
+{
+    int status = xpmem_detach(((struct acoll_xpmem_rcache_reg_t *) reg)->xpmem_vaddr);
+    return status;
+}
+#endif
+
+static inline int coll_acoll_init(mca_coll_base_module_t *module, ompi_communicator_t *comm,
+                                  coll_acoll_data_t *data)
+{
+    int size, ret = 0, rank, line;
+
+    mca_coll_acoll_module_t *acoll_module = (mca_coll_acoll_module_t *) module;
+    coll_acoll_subcomms_t *subc;
+    int cid = ompi_comm_get_local_cid(comm);
+    subc = &acoll_module->subc[cid];
```

Comment:
Thanks, will do. Do you foresee any issues with this approach?

Reply:
In theory CID could be a random and large value. Maybe the current implementation somehow guarantees that it's a zero-indexed continuous value but I'm not sure if that will change.

Curious if you see other places in ompi where CID is used as the array index?

Reply:
what's the cid going to be used for?

Reply:
the cid max val is dependent on the PML /MTL being used.  search the code base for ```pml_max_contextid``` and ```mtl_max_contextid```.   Since we usually want to be able to fit the bits needed for message matching in a 64bit quantity, (cid, rank, tag) it can't be that big but its not going to be small either.

Reply:
@wenduwan @hppritcha  we are using it to index into an array of sub-communicators. However, we have a max value (MCA_COLL_ACOLL_MAX_CID) beyond which we don't use cid.  After searching I found the osc_rdma component uses it as a key within a hashtable, but that's all I could find.

We will explore an alternate approach as well (mostly as a separate PR)

Reply:
> @wenduwan @hppritcha we are using it to index into an array of sub-communicators. However, we have a max value (MCA_COLL_ACOLL_MAX_CID) beyond which we don't use cid. After searching I found the osc_rdma component uses it as a key within a hashtable, but that's all I could find.
> 
> We will explore an alternate approach as well (mostly as a separate PR)

Additionally, we use cid to distinguish between the cases where the collective is called using a sub-communicator from another collective. We will explore utilizing cid to distinguish between these cases without it being used as an index in a separate PR.